### PR TITLE
Update Encryption.php

### LINKED ISSUE #9792: Encrypted external WebDAV storage not working in NC 13.0.3, OCP\Encryption\Exceptions\GenericEncryptionException: Missing Signature
### Steps to reproduce
1. Enable server side encryption module (Default encryption module)
2. Leave all settings at default (external storage encrypted, local storage not encrypted)
3. Mount an external WebDAV folder using a normal user, allow sharing for that folder. In my example WebDAV provider is https://webdav.magentacloud.de/

### Expected behaviour
Folder should work

### Actual behaviour
Upload to the remote folder works, download doesnt work. NC's Error log has lot of entries stating
OCP\Encryption\Exceptions\GenericEncryptionException: Missing Signature (complete error message will be at the end)
Looking directly at the WebDAV storage (using an WebDAV client) reveals that the file was uploaded successfully and is encrypted, but there are no hidden signature files

### Server configuration

**Operating system**:
Linux 2.6.32 x86_64

**Web server:**
Apache 2.2

**Database:**
MySQL 5.1.73

**PHP version:**
7.0.29

**Nextcloud version:** (see Nextcloud admin page)
13.0.3 (same issue with 13.0.2)

**Updated from an older Nextcloud/ownCloud or fresh install:**
Updated from 12.x

**Where did you install Nextcloud from:**
Web based installer https://download.nextcloud.com/server/installer/setup-nextcloud.php

**Signing status:**

<details>
<summary>No errors have been found.</summary>

```
Login as admin user into your Nextcloud and access 
http://example.com/index.php/settings/integrity/failed 
paste the results here.
```
</details>

**List of activated apps:**
<details>
<summary>All default + encryption module</summary>

```
If you have access to your command line run e.g.:
sudo -u www-data php occ app:list
from within your Nextcloud installation folder
```
No sudo (shared host)
</details>

Detailled error message:
	OCP\Encryption\Exceptions\GenericEncryptionException: Missing Signature

    /var/www/virtual/<path>/apps/encryption/lib/Crypto/Crypt.php - line 529: OCA\Encryption\Crypto\Crypt->hasSignature('---------------...', 'AES-256-CTR')
    /var/www/virtual/<path>/apps/encryption/lib/Crypto/Crypt.php - line 462: OCA\Encryption\Crypto\Crypt->splitMetaData('---------------...', 'AES-256-CTR')
    /var/www/virtual/<path>/apps/encryption/lib/Crypto/Encryption.php - line 380: OCA\Encryption\Crypto\Crypt->symmetricDecryptFileContent('---------------...', 'D\x08!\x8E\x99\x99$\x08`fw\xA8\xDF\xA0H...', 'AES-256-CTR', 1, 0)
    /var/www/virtual/<path>/lib/private/Files/Stream/Encryption.php - line 464: OCA\Encryption\Crypto\Encryption->decrypt(*** sensitive parameters replaced ***)
    /var/www/virtual/<path>/lib/private/Files/Stream/Encryption.php - line 295: OC\Files\Stream\Encryption->readCache()
    [internal function] OC\Files\Stream\Encryption->stream_read(8192)
    /var/www/virtual/<path>/apps/files_external/3rdparty/icewind/streams/src/Wrapper.php - line 83: fread(Resource id #55, 8192)
    /var/www/virtual/<path>/apps/files_external/3rdparty/icewind/streams/src/CallbackWrapper.php - line 91: Icewind\Streams\Wrapper->stream_read(8192)
    [internal function] Icewind\Streams\CallbackWrapper->stream_read(8192)
    /var/www/virtual/<path>/3rdparty/sabre/http/lib/Sapi.php - line 80: stream_copy_to_stream(Resource id #58, Resource id #59, '5961980')
    /var/www/virtual/<path>/3rdparty/sabre/dav/lib/DAV/Server.php - line 498: Sabre\HTTP\Sapi sendResponse(Object(Sabre\HTTP\Response))
    /var/www/virtual/<path>/3rdparty/sabre/dav/lib/DAV/Server.php - line 254: Sabre\DAV\Server->invokeMethod(Object(Sabre\HTTP\Request), Object(Sabre\HTTP\Response))
    /var/www/virtual/<path>/apps/dav/lib/Server.php - line 287: Sabre\DAV\Server->exec()
    /var/www/virtual/<path>/apps/dav/appinfo/v2/remote.php - line 35: OCA\DAV\Server->exec()
    /var/www/virtual/<path>/remote.php - line 164: require_once('/var/www/virtua...')
    {main}

### PR DESCRIPTION
**Pull request by martink-p**
Fix #9792 
Pull request for a working solution of issue #9792 (https://github.com/nextcloud/server/issues/9792)

### CODE REVIEW

File: lib/private/Files/Stream/Encryption.php [hunk 1]
```diff
@@ -315,6 +315,23 @@ public function stream_read($count) {
 		return $result;
 
 	}
+	
+	/**
+	* stream_read wrapper to read complete requested block
```
Comment (kesselb) [inline lib/private/Files/Stream/Encryption.php]: Thank you very much :+1: You explained in issue (or the owncloud issue not sure) why this change is required. Would you mind to add some technical explanation here as comment? Because of X we need Y and that is done by this method. That makes it easier for other people to follow the code.

File: lib/private/Files/Stream/Encryption.php [hunk 2]
```diff
@@ -315,6 +315,23 @@ public function stream_read($count) {
...
+	*/
+	private function stream_read_block($blockSize) {
```
Comment (kesselb) [inline lib/private/Files/Stream/Encryption.php]:
```suggestion
	private function stream_read_block(int $blockSize) {
```

Should be save to typehint $blockSize

File: lib/private/Files/Stream/Encryption.php [hunk 3]
```diff
@@ -315,6 +315,23 @@ public function stream_read($count) {
...
+		$remaining = $blockSize;
+		$data = "";
```
Comment (kesselb) [inline lib/private/Files/Stream/Encryption.php]:
```suggestion
		$data = '';
```

File: lib/private/Files/Stream/Encryption.php [hunk 4]
```diff
@@ -315,6 +315,23 @@ public function stream_read($count) {
...
+
+		do {
+			$chunk = parent::stream_read($remaining);
+			$chunk_len = strlen($chunk);
+			$data .= $chunk;
+			$remaining -= $chunk_len;
+		} while ( ($remaining > 0) && ($chunk_len > 0) );
```
Comment (kesselb) [inline lib/private/Files/Stream/Encryption.php]:
```suggestion
		} while (($remaining > 0) && ($chunk_len > 0));
```

File: lib/private/Files/Stream/Encryption.php (hunk 1 above)
Comment (kesselb) [inline lib/private/Files/Stream/Encryption.php]:
```suggestion
	 * stream_read wrapper to read complete requested block
```

File: lib/private/Files/Stream/Encryption.php [hunk 5]
```diff
@@ -315,6 +315,23 @@ public function stream_read($count) {
 		return $result;
 
 	}
+	
+	/**
+	* stream_read wrapper to read complete requested block
+	*/
```
Comment (kesselb) [inline lib/private/Files/Stream/Encryption.php]:
```suggestion
	 */
```

**Comment by kesselb** [comment]
Ref https://github.com/owncloud/core/issues/34599 and https://github.com/owncloud/core/pull/35519 for some more details

File: lib/private/Files/Stream/Encryption.php (hunk 1 above)
Comment (martink-p) [reply lib/private/Files/Stream/Encryption.php]: Yes of course I can:
To my findings the underlying file stream does not return a constant block size on read. It does return as much as it can on every call. This is probably because the stream receives data over the internet asynchroneously when you use and exteranal (e.g. dav) storage. Additionally it seems that stream_read (or subsequent functions) does (do) not care about the requested block size you pass.
Thats why I've added a loop which calls stream_read over again until the requested block size is reached or there is no remaining data.

Comment (kesselb) [reply lib/private/Files/Stream/Encryption.php]: Thank you ;) I had in mind to add these explanation to the source code ;) 

```
/*
 * stream_read wrapper to read complete requested block
 *
 * Because of X we need Y and that is done by this method.
 */
```

Comment (martink-p) [reply lib/private/Files/Stream/Encryption.php]: Sure ;) Obviously I'm a bit slow-witted today ;)

Comment (kesselb) [reply lib/private/Files/Stream/Encryption.php]: Cool :+1: Just another small nitpick :see_no_evil: 

```
	/**
	 * stream_read_block
	 *
	 * This function is a wrapper for function stream_read.
	 * It calls stream read until the requested $blockSize was received or no remaining data is present.
	 * This is required as stream_read only returns smaller chunks of data when the stream fetches from a
	 * remote storage over the internet and it does not care about the given $blockSize.
	 *
	 * @param int $blockSize Length of requested data block in bytes
	 * @return string Data fetched from stream.
	 */
```

In the php world we use https://phpdoc.org/ for documenting code.

Comment (martink-p) [reply lib/private/Files/Stream/Encryption.php]: Sure, if forgot. I'll change this in a minute.

**Review by rullzer — APPROVED** [review]

**Review by schiessle — APPROVED** [review]
@martink-p thanks for your pull request! Looks good to me and makes absolutely sense.
